### PR TITLE
Use Typescript 5.1 JSX ElementType

### PR DESCRIPTION
**Pull request by petersalas**
This uses the [new Typescript 5.1 feature](https://devblogs.microsoft.com/typescript/announcing-typescript-5-1/#decoupled-type-checking-between-jsx-elements-and-jsx-tag-types) to add proper type-checking for AI.JSX elements.

Additionally it fixes a couple places where were (correctly!) identified as incorrectly typed.

### CODE REVIEW

**Comment by petersalas** [comment]
@NickHeiner how do you feel about requiring TS 5.1 for our types? On one hand it's a barrier to adoption, OTOH it's one that will go down (presumably quickly) over time. Is the TS community generally very quick to adopt new TS versions?

File: packages/examples/src/debate.tsx [hunk 1]
```diff
@@ -27,32 +27,29 @@ async function ConversationRemapper(
...
-      }
+    if (assistant(node)) {
+      return <AssistantMessage>{node}</AssistantMessage>;
+    }
 
-      return node;
-    })
-    .filter((n) => n !== null);
+    return node;
+  });
```
Comment (NickHeiner) [inline packages/examples/src/debate.tsx:52]: Is this just a formatting change?

**Comment by NickHeiner** [comment]
I was hoping this would also clean up some of the types in NextJS, but I recognize that's a hot mess.

**Review by NickHeiner — APPROVED** [review]

File: packages/examples/src/debate.tsx (hunk 1 above)
Comment (petersalas) [reply packages/examples/src/debate.tsx:52]: Not quite, the types now allowed TS-lint to correctly identify that the `filter` was not necessary. (Which apparently triggered prettier to reformat the rest of the code.)

**Comment by petersalas** [comment]
> I was hoping this would also clean up some of the types in NextJS, but I recognize that's a hot mess.

I think we can likely use the same strategy to provide different separate JSX types for the integration.

**Pull request merged**